File: dotnet/Allmark/Inline/LineBreakRule.cs
namespace Allmark.Inline;

using Allmark.Types;

public static class LineBreakRule
{
	public static InlineRule Create()
	{
		return new InlineRule
		{
			Name = "line_break",
			Test = TestLineBreak,
		};
	}

	private static bool TestLineBreak(InlineParserState state, MarkdownNode parent)
	{
		var ch = Utils.GetChar(state.Src, state.I);
		if (ch == ' ')
		{
			var end = state.I;
			for (var i = state.I + 1; i < state.Src.Length; i++)
			{
				if (Utils.IsNewLine(state.Src[i]))
				{
					end = i;
					break;
				}
				else if (state.Src[i] == ' ')
				{
					continue;
				}
				else
				{
					return false;
				}
			}
			if (end - state.I >= 2)
			{
				var html = Utils.NewNode("html_span", false, state.I, state.Line, 1, "", state.Indent);
				html.Content += "<br />\n";
				parent.Children!.Add(html);
				state.I = end + 1;
				return true;
			}
		}

		return false;
	}
}

File: zig/src/inline/lineBreakRule.zig
const std = @import("std");
const InlineParserState = @import("../types/InlineParserState.zig").InlineParserState;
const InlineRule = @import("../types/InlineRule.zig").InlineRule;
const MarkdownNode = @import("../types/MarkdownNode.zig").MarkdownNode;
const isNewLine = @import("../utils/isNewLine.zig").isNewLine;
const newNode = @import("../utils/newNode.zig").newNode;

pub fn testLineBreak(state: *InlineParserState, parent: *MarkdownNode) bool {
    const char = if (state.i < state.src.len) state.src[state.i] else 0;
    if (char == ' ') {
        var end = state.i;
        var i = state.i + 1;
        while (i < state.src.len) {
            if (isNewLine(&.{state.src[i]})) {
                end = i;
                break;
            } else if (state.src[i] == ' ') {
                i += 1;
                continue;
            } else {
                return false;
            }
            i += 1;
        }

        if (end - state.i >= 2) {
            const html = newNode(state.allocator, "html_span", false, state.i, state.line, 1, "", state.indent, null) catch unreachable;
            html.*.content = "<br />\n";
            const children = parent.children orelse {
                parent.children = std.ArrayList(*MarkdownNode).initCapacity(state.allocator, 1) catch unreachable;
                parent.children.?.appendAssumeCapacity(html);
                state.i = end + 1;
                return true;
            };
            children.append(html) catch unreachable;
            state.i = end + 1;
            return true;
        }
    }

    return false;
}

pub const lineBreakRule = InlineRule{
    .name = "line_break",
    .@"test" = testLineBreak,
};
